cmd/codegen/generator/go/templates/modules.go

```diff
@@ -677,99 +716,190 @@ func (ps *parseState) goMethodToAPIFunctionDef(typeName string, fn *types.Func,
-						if param.Type().String() == "string" {
-							enc, err := json.Marshal(tag.Value())
-							if err != nil {
-								return nil, fmt.Errorf("failed to marshal default value: %w", err)
-							}
-							jsonEnc = string(enc)
-						} else {
-							jsonEnc = tag.Value() // assume JSON encoded
-						}
-						argOpts = append(argOpts, Id("DefaultValue").Op(":").Id("JSON").Call(Lit(jsonEnc)))
-					}
-				}
+		i++
+	}
 
-				if len(argOpts) > 0 {
-					argArgs = append(argArgs, Id("FunctionWithArgOpts").Values(argOpts...))
-				}
+	fnDecl, err := ps.declForFunc(fn)
+	if err != nil {
+		return nil, err
+	}
 
-				fnDef = dotLine(fnDef, "WithArg").Call(argArgs...)
+	if i+1 == params.Len() {
+		param := params.At(i)
+		paramType, ok := asInlineStruct(param.Type())
+		if ok {
+			stype, ok := asInlineStructAst(fnDecl.Type.Params.List[i].Type)
+			if !ok {
+				return nil, fmt.Errorf("expected struct type for %s", param.Name())
 			}
-		} else {
-			argTypeDef, err := ps.goTypeToAPIType(param.Type(), nil)
-			if err != nil {
-				return nil, fmt.Errorf("failed to convert param type: %w", err)
+
+			parent := &paramSpec{
+				name:      params.At(i).Name(),
+				paramType: param.Type(),
+				baseType:  param.Type(),
 			}
 
-			fnDef = dotLine(fnDef, "WithArg").Call(Lit(param.Name()), argTypeDef)
+			for f := 0; f < paramType.NumFields(); f++ {
+				spec, err := ps.parseParamSpecVar(paramType.Field(f))
+				if err != nil {
+					return nil, err
+				}
+				spec.parent = parent
+				spec.description = stype.Fields.List[f].Doc.Text()
+				if spec.description == "" {
+					spec.description = stype.Fields.List[f].Comment.Text()
+				}
+				spec.description = strings.TrimSpace(spec.description)
+				specs = append(specs, spec)
+			}
+			return specs, nil
 		}
 	}
 
-	return fnDef, nil
+	for ; i < params.Len(); i++ {
+		spec, err := ps.parseParamSpecVar(params.At(i))
+		if err != nil {
+			return nil, err
+		}
+		if sig.Variadic() && i == params.Len()-1 {
+			spec.variadic = true
+		}
+
+		pos := fnDecl.Type.Params.List[i].Pos()
+		if cmt, err := ps.commentForPos(pos - 1); err == nil {
+			spec.description = cmt.Text()
+			spec.description = strings.TrimSpace(spec.description)
+		}
+
+		specs = append(specs, spec)
+	}
+	return specs, nil
 }
 
-func namedOrDirectStruct(t types.Type) (*types.Struct, bool) {
-	switch x := t.(type) {
-	case *types.Named:
-		return namedOrDirectStruct(x.Underlying())
-	case *types.Struct:
-		return x, true
-	default:
-		return nil, false
+func (ps *parseState) parseParamSpecVar(field *types.Var) (paramSpec, error) {
+	if _, ok := field.Type().(*types.Struct); ok {
+		return paramSpec{}, fmt.Errorf("nested structs are not supported")
+	}
+
+	paramType := field.Type()
+	baseType := paramType
+	for {
+		ptr, ok := baseType.(*types.Pointer)
+		if !ok {
+			break
+		}
+		baseType = ptr.Elem()
+	}
+
+	optional := false
+	if named, ok := baseType.(*types.Named); ok {
+		if named.Obj().Name() == "Optional" {
```

Comment:
nit: may want to also check if `named` is coming from `dagger.gen.go`, just to future-proof against support we add eventually to use types from external packages (which could also be named `Optional` here since `.Obj().Name()` returns an unqualified name)